airflow/jobs.py

```diff
@@ -403,39 +422,77 @@ def schedule_dag(self, dag):
                     dr.end_date = datetime.now()
             session.commit()
 
+            # check last scheduled run
+            # get this dag's max execution date by
+            #  external trigger is false
+            #  and "like" Dag Run ID.ID
+            # bolke: why a like?
             qry = session.query(func.max(DagRun.execution_date)).filter_by(
-                    dag_id = dag.dag_id).filter(
-                        or_(DagRun.external_trigger == False,
-                            # add % as a wildcard for the like query
-                            DagRun.run_id.like(DagRun.ID_PREFIX+'%')))
+                dag_id=dag.dag_id).filter(
+                or_(DagRun.external_trigger is False,
+                    # add % as a wildcard for the like query
+                    DagRun.run_id.like(DagRun.ID_PREFIX+'%')
+                    )
+            )
             last_scheduled_run = qry.scalar()
+            self.logger.debug("Last scheduled run: {}".format(last_scheduled_run))
+
+            # Determine next run date
             next_run_date = None
             if dag.schedule_interval == '@once' and not last_scheduled_run:
                 next_run_date = datetime.now()
             elif not last_scheduled_run:
                 # First run
+                self.logger.debug("Scheduling {} for the first time".format(dag.dag_id))
+
+                # Get the maximum execution date for the tasks associated with this dag
                 TI = models.TaskInstance
                 latest_run = (
                     session.query(func.max(TI.execution_date))
                     .filter_by(dag_id=dag.dag_id)
                     .scalar()
                 )
+
                 if latest_run:
                     # Migrating from previous version
                     # make the past 5 runs active
+                    # todo: remove this?
+                    # fixme: from what version and why not fixed in alembic?
+                    self.logger.warning("Migrating from previous version for dag {} "
+                                        "marking past 5 runs active".format(dag.dag_id))
                     next_run_date = dag.date_range(latest_run, -5)[0]
                 else:
+                    # Set start_date based on tasks
                     task_start_dates = [t.start_date for t in dag.tasks]
                     if task_start_dates:
-                        next_run_date = min(task_start_dates)
```

Comment:
Well, I think (this logic that was already there) that due to the fact that tasks can have their own start_date (and they inherit this from the dag) this can pre-empt the set start date of the dag.

Ie your dag can have a start_date of 2016-04-25, but the tasks contained in the dag set 2016-05-10 as the minimum then the date of the next run will be 2016-05-10. Obviously you want to start when the first task is ready hence min()